File: play-ws-standalone/src/main/scala/play/api/libs/ws/StandaloneWSClient.scala
/*
 * Copyright (C) from 2022 The Play Framework Contributors <https://github.com/playframework>, 2011-2021 Lightbend Inc. <https://www.lightbend.com>
 */

package play.api.libs.ws

import java.io.Closeable
import java.io.IOException

/**
 * The WSClient holds the configuration information needed to build a request, and provides a way to get a request holder.
 */
trait StandaloneWSClient extends Closeable {

  /**
   * The underlying implementation of the client, if any.  You must cast explicitly to the type you want.
   *
   * @tparam T the type you are expecting (i.e. isInstanceOf)
   * @return the backing class.
   */
  def underlying[T]: T

  /**
   * Generates a request.  Throws IllegalArgumentException if the URL is invalid.
   *
   * @param url The base URL to make HTTP requests to.
   * @return a request
   */
  @throws[IllegalArgumentException]
  def url(url: String): StandaloneWSRequest

  /**
   * Closes this client, and releases underlying resources.
   */
  @throws[IOException]
  def close(): Unit
}

File: play-ws-standalone/src/main/java/play/libs/ws/StandaloneWSClient.java
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package play.libs.ws;

import java.io.IOException;

/**
 * This is the WS Client interface.
 */
public interface StandaloneWSClient extends java.io.Closeable {

    /**
     * The underlying implementation of the client, if any.  You must cast the returned value to the type you want.
     *
     * @return the backing class.
     */
    Object getUnderlying();

    /**
     * Returns a StandaloneWSRequest object representing the URL.  You can append additional
     * properties on the StandaloneWSRequest by chaining calls, and execute the request to
     * return an asynchronous {@code CompletionStage<StandaloneWSResponse>}.
     *
     * @param url the URL to request
     * @return the request
     */
    StandaloneWSRequest url(String url);

    /**
     * Closes this client, and releases underlying resources.
     * <p>
     * Use this for manually instantiated clients.
     */
    void close() throws IOException;
}
